### DISCUSSION TITLE
How to generate the required TOPAS parameter files

### DISCUSSION THREAD

**Discussion by Yuri-HubL** [e0404/matRad#636] [Q&A] [answered]
Hi matRad Expert

Now I am trying to use dev_varRBErobOpt, to generate the TOPAS parameter file, I see example is matRad_example12_simpleParticleMonteCarlo m files.

I set the following command according to the prompt, but it doesn't seem to record any TOPAS parameters, which confuses me.
**pln.propMC.externalCalculation = true;**
I attach my m file to the back and expect someone to point out my errors and tell me how to generate the required TOPAS parameter file[my_example12.txt](https://github.com/e0404/matRad/files/11812241/my_example12.txt)

your sincerely;
Yuri

**Comment by wahln** [comment]
If no error occured, parameter files will be created in the subfolder topas/MCrun.

**Comment by remocristoforetti** [comment]
Hello Yuri,
thank you for your question and for sharing your script.
I took a look at it and I noticed some minor issues in the setup that can be easily solved.

Here is a list of changes that you can introduce to make this matRad script work:

- ` pln.multScen` is a field of the `pln` structure that has to be always defined. This property is used to define and apply robustness techniques during dose calculation and fluence optimization. If, as in this case, no robustness technique needs to be applied, this field should be instantiated as follows:

   `pln.multScen = matRad_multScen(ct,'nomScen')`

   This will just include the nominal "scenario", i.e. no robustness technique will be applied.
   As you can notice, this line of code is already present in your script, it only needs to be de-commented.

- another important field of the `pln` struct that has to be defined is the `pln.bioParam` field. This allows for the application of the specified biological model for RBE calculation. If you are only interested in the physical dose distribution, please set this field as follows:
```
   %define the quantity to use for calculation and optimization
   quantityOpt = 'physicalDose';
   %define the model to be used
   model = 'none';
   pln.bioParam = matRad_bioModel(pln.radiationMode,quantityOpt,model);
```
   This will disable the RBE distribution calculation and optimization.

- Setting:  `pln.propMC.externalCalculation = true;` as you did is indeed the correct way to allow matRad to produce the TOPAS parameters files as an output. However, I noticed that in your script the line:
`resultGUI_MC = matRad_calcDoseDirectMC(ct,stf,pln,cst,resultGUI.w);`
seems to be commented. Please run this function in order to correctly trigger the Monte Carlo dose calculation (and thus the generation of the TOPAS parameter files).

When correctly run, matRad will save the parameter files in the directory ` topas/MCrun/`

Please note that the **dev_varRBErobOpt** branch is still under development and has not yet been fully tested. Updates are indeed continuously provided.
Please also note that we recently added some minor fixes to this branch, and more fixes will come in the near future. I advise you keep your version of the code updated.

Please find attached a modified version of your script that includes the above mentioned changes.

I remain at your disposal for any further clarification.
Thank you for using matRad!

Best,
Remo
[my_example12.txt](https://github.com/e0404/matRad/files/11816188/my_example12.txt)

**Comment by Yuri-HubL** [comment]
Hi Remo，

Thank you for your reply.

- According to your suggestions, I have successfully modified my files and can successfully generate TOPAS files. However, this seems to be limited to "proton". I will attach the files that I successfully run to the rear.  "proton.txt"

- But if I make a small change to this file,

```
pln.radiationMode   = 'carbon';
pln.machine         = 'Generic';
pln.numOfFractions  = 30;
```
Then there will be some errors, I don't know how to solve them

![wrong](https://github.com/e0404/matRad/assets/72288469/500c65a9-8e8f-4977-891c-8af7d5611fd9)
[carbon.txt](https://github.com/e0404/matRad/files/11858519/carbon.txt)
[proton.txt](https://github.com/e0404/matRad/files/11858520/proton.txt)

Looking forward your relpy.
your sincerely,
Yuri

**Comment by remocristoforetti** [comment]
Hi Yuri,
I had a look at the carbon script, and I notice that you are trying to use the TOPAS interface to compute the `dij` variable in **matRad** through the function `matRad_calcParticleDoseMC`.
Sadly, this functionality has not yet been tested, nor officialy made available in this version of **matRad**. We are, indeed, still working on this.

However, I suggest you apply the same approach you correctly setup for the proton case. 
So first compute the `dij` variable through the analytical approach (via `matRad_calcParticleDose`, not `matRad_calcParticleDoseMC`), then optimize the fluence weights and only after that, perform the re-calculation of the plan with **TOPAS**.

Let me know if you have any further question,
Best regards,
Remo

P.S. I notice from the proton script that you are trying to setup an RBE weighted dose optimization. This is totally possible in **matRad**, but the '`LEM`' model is actually only available for carbon ions, and not for protons.
This does not output any error since the biological model is recognized as incorrect but substituted with the default one (which is constant RBE, `constRBE`, for protons). Please keep in mind this feature when analyzing your results.

**Comment by Yuri-HubL** [comment]
Hi Remo,

Thank you very much for your reply and suggestions

I did use the matRad_calcParticleDose function to calculate the dij of the carbon ion, which would give a good weight, but matlab would give an error saying "The sample points must be finite".

![sample points is unfinite](https://github.com/e0404/matRad/assets/72288469/1c1038ac-9f37-4b9c-80f4-47a9b0b51576)

This question bothers me too. Looking forward to your suggestions.

your sincerely,

Yuri

**Reply by remocristoforetti** [reply]
Hi Yuri,
looks like an interpolation issue.
Could you please share the exact script that generated the error? I was not able to reproduce this with the scripts that you sent before, which should now work correctly.
Best,
Remo

**Reply by Yuri-HubL** [reply]
Hi Remo，
This is my new executable file code.I only changed one line of code.
`dij = matRad_calcParticleDose(ct,stf,pln,cst);`
The following error appears during the run.
```
Error using griddedInterpolant
The sample points must be finite.
```
[carbon_new.txt](https://github.com/e0404/matRad/files/11875451/carbon_new.txt)

I hope you can give me some advice. Looking forward to your reply.

your sincerely,
Yuri

**Reply by remocristoforetti** [reply]
Hi Yuri,
thank you for sharing your scripts.
I tried to run this new script and it does not produce any error on my branch. Are you sure your version of the code is up to date with the dev_varRBErobOpt branch?
These functionalities and the matRad-TOPAS interface are actually under continuous development, it might be that this bug has already been fixed with one of the latest updates.

Have a nice day,
Remo

**Reply by Yuri-HubL** [reply]
Hi Remo,
Thank you for your reply. I checked my dev_varRBErobOpt branch and it was indeed the latest version.
Is this due to the incompatibility of the matlab version? , my matlab version is 2021b.
your sincerely，
Yuri

**Reply by remocristoforetti** [reply, marked as answer]
Hi Yuri,
the function that gives that error message seems to be in the class matRad_MCEmittanceBaseData. Please check that particularly lines `502-503` in that function are up to date. One of the last commits ([https://github.com/e0404/matRad/commit/99eaa34bff3b9089dcefd6d6353735e3e4bfec05](url)) on the branch addresses precisely those lines.

Please also check that the whole function is updated, it has undergone some fixes in the past days.
Let me know if this solves the issue,
best
Remo

**Reply by Yuri-HubL** [reply]
Dear Remo,

I updated to the latest dev_varRBErobOpt branch, which successfully solved my problem!
Thank you very much!

your sincerely,
Yuri

**Reply by remocristoforetti** [reply]
Thank you for using matRad.
Best,
Remo

**Comment by Yuri-HubL** [comment]
Hi Remo,

I am sending this email to ask you a new question. After successfully executing the code, I got a matRad_cube.dat, and I wanted to know what tool I could use to read the information in this format file.

So far I've tried a few plain text readers like notepad and I've also tried visual studio but it doesn't work.

Sorry to bother you again, looking forward to your reply.

your sincerely,
Yuri

**Reply by Yuri-HubL** [reply]
Hi all，

I seem to have found the answer, and I hope this will be useful to all user as well.
https://github.com/davidchall/topas2numpy
https://topas.readthedocs.io/en/latest/examples-docs/Patient/DoseToCT.html?highlight=xio%20dat
If you have more methods welcome to add.
Hope you have a good day.

your sincerely,
Yuri

**Reply by wahln** [reply]
The matRad_cube.dat is, afaik, just the CT from matRad in a format that TOPAS can read it to perform the simulation - it is used within TOPAS as "TsImageCube" (which you can see in matRad_cube.txt)

**Reply by Yuri-HubL** [reply]
Hi Wahl,

Thank you very much for your reply. I have a question I looked at the folder matRad_cube.txt and I think I should understand the meaning in the file.

1. The first is DensityCorrection: This is the correction factor that is required for all HU values to be converted into densities.
2. The second SchneiderHounsfieldUnitSections, SchneiderHounsfieldUnitSections, SchneiderDensityFactor, SchneiderDensityFactorOffset , SchneiderHUToMaterialSections: According to the personalized generation of specific patient information, the specific items in the transformation formula are specified：
Density = (Offset + (Factor * (FactorOffset + HU[-1000,2995]))) * DensityCorrection
![Schneider](https://github.com/e0404/matRad/assets/72288469/da7a8b55-d8ed-48a7-8fc1-78bd1cc59e7e)

3. SchneiderElements, SchneiderMaterialsWeight1, SchneiderMaterialsWeight2, SchneiderMaterialMeanExcitationEnergy: The information of the patient tissue material is defined.
![materials](https://github.com/e0404/matRad/assets/72288469/319d4114-a520-4e33-9c78-b1a3c9664b16)

4. patient paramete: where matRad_cube.dat seems to record exactly the numerical information in each voxel of the patient.
![patient parameter](https://github.com/e0404/matRad/assets/72288469/5e60ccfc-1559-4e2d-969b-e213e4b54a36)
[matRad_cube.txt](https://github.com/e0404/matRad/files/11895403/matRad_cube.txt)

So I think the information recorded in matRad_cube.txt and matRad_cube.dat is different.Is my understanding correct?

And I concern is how can 3 d visualization matRad_cube. The information in the dat, now I am trying to modify the topas/example/patient/DoseToCT. txt to visualize matRad generated in the file.

I will continue to ask questions if I have any further doubts. Thank you again for your reply. Nice talking to you

your sincerely,
Yuri

**Reply by remocristoforetti** [reply]
Hi Yuri,
it is correct that the information in `matRad_cube.txt` and `matRad_cube.dat` is different. The `.txt` file is the parameter file that collects the **TOPAS** commands to create and place the `/Patient` component, while the `.dat` file just contains the HU for every voxel.

The `.dat` cube is nothing else than the field entry `ct.cubeHU` from **matRad**. If you want to visualize this cube, and since you are already familiar with **Matlab**, I suggest you exploit the plotting capabilities of **Matlab** itself to visualize the data. (You can also quickly load the `.dat` file via the function "`fread`" provided by **Matlab**).

Alternatively, you can have a look at the plotting tools provided by **matRad** (that you can find in `/plotting` directory).

Have a nice day,
Remo

**Reply by Yuri-HubL** [reply]
Hi Remo,
Thank you very much for your reply, it helped me a lot!

It's been a pleasure talking to you. Thank you for developing and maintaining matRad, and thank you for being so helpful in answering users' questions.

Have a good day,
Yuri
